building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
++ add_library(sse "${SRC_FILES}")
+- add_executable("${PROJECT_NAME}" "${SRC_FILES}")
+```
+
+- Build static lib `libHelloWorld.a`, that's the default
+- Build shared lib by `cmake -DBUILD_SHARED_LIBS=ON ..`, don't change in `CMakeLists.txt`, but stick to standards
+- Define cmake variables with `-D`, we will come back to this later
+- Now, let's use the library
+
+```diff
++ add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- ... gives an error, we also need to link
+
+```diff
++ target_link_libraries("${PROJECT_NAME}" PUBLIC sse)
+```
+
+TODO: explain PRIVATE (only build), PUBLIC (both), INTERFACE (only linking)
+
+### External Dependencies
+
+- preCICE should be used in every code :)
+
+`main.cpp`:
+
+```diff
++ #include "precice/SolverInterface.hpp"
+...
++ sse();
++ std::cout << precice::getVersionInformation() << std::endl;
+```
+
+`CMakeLists.txt`:
+
+```diff
+find_package(precice REQUIRED CONFIG)
+target_link_libraries("${PROJECT_NAME}" PRIVATE precice::precice)
+```
+
+- `find_package` works if dependency is "cmake-ready"
+- Alternative option if lib is in a discoverable location (standard system path or adjusted `LD_LIBRARY_PATH`)
+
+```cmake
+find_library(precice REQUIRED CONFIG)
+target_link_libraries("${PROJECT_NAME}" PRIVATE precice)
+```
+
+- TODO: external dependencies via pkg-config?
+    - @fsimonis: Is `find_package` already using pkg-config? Or is there a real third option?
+    - include(..), pkg-config.search_modules ... link ... cmake searches in pkg-config
+
+### Options and variables
+
+- For good software, a user should not need to touch `CMakeLists.txt`
+- Use variables to talk to cmake
+- Let's try to make preCICE an optional dependency of our code (not really everybody has preCICE installed, sadly)
+
+`CMakeLists.txt`:
+
+```cmake
+if (ENABLE_PRECICE)
+  message(STATUS "preCICE enabled")
+  find_library(precice REQUIRED CONFIG)
+  target_link_libraries("${PROJECT_NAME}" PRIVATE precice)
+else()
+  message(STATUS "preCICE disabled")
+  target_compile_definitions("HelloWorld" PRIVATE NO_PRECICE)
+endif()
+```
+
+`main.cpp`:
+
+```c++
+#ifndef NO_PRECICE
+  std::cout << precice::getVersionInformation() << std::endl;
+#endif
+```
+
+- Build without preCICE `cmake -DENABLE_PRECICE=OFF ..`
+- Create different folder and build there with preCICE
+
+TODO: variables, which ones are import? how to pass? change compiler etc
+
+`CXX ?= g++`
+
+TODO: More reading on this:
+
+- https://cmake.org/cmake/help/latest/module/FindPkgConfig.html#command:pkg_check_modules
+- https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html
+- https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html
+
+
+### ccmake
+
+`sudo apt-get install cmake-curses-gui`
+
+- `ccmake ..`
+- `c`, look add description of ENABLE_PRECICE`, disable it
+- `t`, show all options, helpful to find out which variables exist
+- `g`, generate `Makefile`
+
+TODO: watch some tutorial, learn more
+
+## Explain preCICE `CMakeLists.txt`
```

Comment:
Instead, maybe introduce "limitations of CMake".

There are some issues as CMake aims to be the common denominator of build systems.
Example: Globs in CMake -> files in build system: Adding new file -> not respected by build system.

Reply:
I would support such a section. (GNU) Make was treated so badly. No it is time for revenge. :smile: 

More seriously: It makes sense to mention the shortcomings/pitfalls when using CMake. It is not perfect and might give trouble in certain settings. The students should be made aware of such situations.

Reply:
I don't think I have the expertise right now to do such a section.
I am also not sure whether we really need this. In the end, our learning goal here is not that students can compare different build system and pick the best one for a certain project. The learning goal is that they understand how cmake roughly works since it used in many simulation codes.

> Example: Globs in CMake -> files in build system: Adding new file -> not respected by build system.

This one I already mention implicitly above